source/agora/network/NetworkManager.d

```diff
@@ -57,31 +57,288 @@ mixin AddLogger!();
+
+        Note that this is a never-ending task. A node's NodeInfo may signal
+        that it's complete, however the list of its peers may grow as new
+        incoming connections are established as well as when any quorum
+        reshuffling happens.
+
+    ***************************************************************************/
+
+    private class AddressDiscoveryTask
+    {
+        import std.container : DList;
+
+        /// A queue of clients. Each client is contacted, and pushed back to the
+        /// queue (unless they were banned)
+        private DList!NetworkClient clients;
+
+        /// Called when we have retrieved a set of addresses from a client
+        private void delegate (Set!Address addresses) onNewAddresses;
+
+
+        /***********************************************************************
+
+            Ctor
+
+            Params:
+                onNewAddresses = called when a set of addresses were retrieved
+                                 from a node
+
+        ***********************************************************************/
+
+        public this (void delegate (Set!Address addresses) onNewAddresses )
+        {
+            this.onNewAddresses = onNewAddresses;
+        }
+
+        /***********************************************************************
+
+            Start the asynchronous address discovery task.
+
+        ***********************************************************************/
+
+        public void run ()
+        {
+            // workaround to only run this task once
+            // (cannot run it in the ctor as the scheduler may not be running yet)
+            // todo: move it to the ctor once we have `schedule` implemented.
+            static bool is_running;
+            if (!is_running)
+            {
+                is_running = true;
+                this.outer.taskman.runTask(&this.getNewAddresses);
+            }
+        }
+
+        /***********************************************************************
+
+            Add a new client to the list of clients to query addresses from.
+
+        ***********************************************************************/
+
+        public void add (NetworkClient client)
+        {
+            this.clients.insertBack(client);
+        }
+
+        /***********************************************************************
+
+            Neverending function that retrieves the network state out of
+            all the clients it knows about. The clients are kept in a queue,
+            and each client is queried in sequence.
+
+            If the client's address is banned, it will be removed from the
+            clients list.
+
+        ***********************************************************************/
+
+        private void getNewAddresses ()
+        {
+            while (1)
+            {
+                scope (success)
+                    this.outer.taskman.wait(1.seconds);
```

Comment:
Ditto why scope success ?

Reply:
In this specific case it's because of the `continue` below. I didn't want to duplicate the wait call.